### DISCUSSION TITLE
classes/class_animationplayer

### DISCUSSION THREAD

**Discussion by giscus** [godotengine/godot-docs-user-notes#181] [User-contributed notes]
# classes/class_animationplayer

Inherits: AnimationMixer< Node< Object A node used for animation playback. Description: An animation player is used for general-purpose playback of animations. It contains a dictionary of Animation...

https://docs.godotengine.org/en/stable/classes/class_animationplayer.html

**Comment by the-alex-devs** [comment]
If you need the .play() function to update immediately, as mentioned in the method description on this page, you'll need to call AnimationPlayer.advance(0). To do this, it must be in a different line than the play() function. For example:
```gdscript
AnimationPlayer.play("HitAnimation")
AnimationPlayer.advance(0)
```

**Comment by pierogigon** [comment]
There is a method that is not listed here called:

get_animation_list()

The above method returns a list of all of the animations an AnimationPlayer has.

**Reply by Calinou** [reply]
`get_animation_list()` is part of AnimationMixer, which is a base class shared by AnimationPlayer and AnimationTree. This is why it doesn't appear in the AnimaionPlayer class reference, but in the AnimationMixer class reference instead.

**Comment by markopolojorgensen** [comment]
`animation_finished` is not emitted when using `play_backwards`.

**Comment by leonardoraele** [comment]
The documentation is a little vague in some parts, so I made some tests...

### ▶ Playing A Non-Looping Animation

If an animation with looping turned off is played, then:

- Signal `current_animation_changed` is emitted twice: once at the start of the animation, *before* `animation_started`, and once at the end, *after* `animation_finished`. The second one has an empty string for argument.
- Signals `animation_started` and `animation_finished` are emitted once each.

### 🔁 Playing A Looping Animation

If an animation with looping turned on is played, then:

- Signals `current_animation_changed` and `animation_started` are emitted once each; and
- Signal `animation_finished` is never emitted.

### ⏭ Playing Another Animation

If an animation is played while the AnimationPlayer is already playing a different animation, then:

- Signal `current_animation_changed` will be emitted once for each played animation, when they start playing, and one more time at the end of the second animation.
- Signal `animation_started` will be emitted once for each played animation.
- Signal `animation_finished` will be emitted only once when the last animation ends, assuming it doesn't have looping enabled.

### ↩ Replaying the Same Animation

If an animation is played with `play()` while the AnimationPlayer is already playing that same animation, then:

- The animation continues playing. It is not interrupted or restarted, but the speed scale is changed to the one used in the second `play()` call, if it's different.
- Contrary to what's in the documentation, the playback *can* be inverted in the middle of the playing animation if the `from_end` parameter is `true` and the `custom_speed` is negative, or if `play_backwards()` is called. You can even change the speed scale inverting the playback like this. Signals `current_animation_changed` and `animation_finished` will both be emitted when the animation completes by playing backwards up to the start.
- The `start_time`, `end_time`, `start_marker`, and `end_marker` parameters of the `play_section_*()` variants seem to have no effect.
- I have not tested the behavior of the `custom_blend` parameter.
- No signals are emitted when the repeated `play()` is called. The initial call to `play()` emits `current_animation_changed` and `animation_started` once each (as normal) when the animation starts playing; and `current_animation_changed` and `animation_finished` are emitted once each when the animation ends.

### 🔽 Queueing Animations

If an animation is played with `play()` and additional animations are queued with `queue()`, then:
- Signal `animation_started` will be emitted once for each animation when they start playing;
- Signal `current_animation_changed` will be emitted once for each animation when they start playing, and one more time, with an empty string argument, when the last one finish;
- Signal `animation_changed` will be emitted once for each queued animation, but not for the initially played animation.
- Signal `animation_finished` will be emitted only once when the last animation finishes.

> ℹ Assuming none of the animations is set to loop.

---

⚠ Note that `animation_started` and `animation_finished` have inconsistent behavior. For example, the former plays when a new `play()` call interrupts an already playing animation, and when queued animations start playing, while the later doesn't in both cases. I don't know if this inconsistency is a bug or not.

**Comment by ottworks** [comment]
current_animation_changed is completely unreliable. do not use it for game logic, just use a timer instead.

**Comment by admin** [comment]
*Originally posted by [**MrYannsen**](https://github.com/MrYannsen) on 2026-06-02 09:40:08 UTC*

> When working with bones and animating characters with symmetrical body parts (left/right arms or legs), it's easy to select the wrong track due to similar naming. It can lead to accidentally modifying finished animations.
>
> Allowing custom colors per track could provide a quick visual distinction and help prevent mistakes.
